src/metadata.rs

```diff
@@ -34,16 +34,40 @@ macro_rules! box_result {
 /// function to fetch metadata. The configured provider is passed in and this
 /// function dispatches the call to the correct provider-specific fetch function
 pub fn fetch_metadata(provider: &str) -> errors::Result<Box<providers::MetadataProvider>> {
+    let legacy = cfg!(feature = "cl-legacy");
+    let fail = Err(errors::ErrorKind::UnknownProvider(provider.to_owned()).into());
+
     match provider {
+        "aws" => {
+            if !legacy {
+                box_result!(AwsProvider::try_new()?)
+            } else {
+                fail
+            }
+        }
         "azure" => box_result!(Azure::try_new()?),
         "cloudstack-metadata" => box_result!(CloudstackNetwork::try_new()?),
         "cloudstack-configdrive" => box_result!(ConfigDrive::try_new()?),
         "digitalocean" => box_result!(DigitalOceanProvider::try_new()?),
-        "ec2" => box_result!(Ec2Provider::try_new()?),
-        "gce" => box_result!(GceProvider::try_new()?),
+        // FIXME: "ec2" is accepted in non-legacy mode temporarily for transition
+        "ec2" => box_result!(AwsProvider::try_new()?),
+        "gce" => {
```

Comment:
Question: doesn't a `#[cfg(feature = "cl-legacy")]` on the match-arm work? (same for all branches)